File: src/spirite/base/image_data/mediums/maglev/MaglevMedium.java
package spirite.base.image_data.mediums.maglev;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import spirite.base.graphics.DynamicImage;
import spirite.base.graphics.GraphicsContext;
import spirite.base.graphics.IImage;
import spirite.base.image_data.ImageWorkspace;
import spirite.base.image_data.ImageWorkspace.BuildingMediumData;
import spirite.base.image_data.layers.puppet.BasePuppet.BaseBone;
import spirite.base.image_data.mediums.ABuiltMediumData;
import spirite.base.image_data.mediums.IMedium;
import spirite.base.image_data.mediums.drawer.IImageDrawer;
import spirite.base.image_data.mediums.maglev.parts.MagLevFill;
import spirite.base.image_data.mediums.maglev.parts.MagLevFill.StrokeSegment;
import spirite.base.image_data.mediums.maglev.parts.MagLevStroke;
import spirite.base.image_data.selection.SelectionMask;
import spirite.base.pen.PenTraits.PenState;
import spirite.base.pen.StrokeEngine.IndexedDrawPoints;
import spirite.base.util.interpolation.Interpolator2D;
import spirite.base.util.linear.MatTrans;
import spirite.base.util.linear.MatTrans.NoninvertableException;
import spirite.base.util.linear.Rect;
import spirite.base.util.linear.Vec2;
import spirite.base.util.linear.Vec2i;
import spirite.hybrid.HybridHelper;

/**
 * A Maglev Internal Image is an image that floats just above the surface, 
 * not quite planting its feat in the ground.  Essentially it is a kind of 
 * Scalable Vector Image, storing all the different stroke and fill actions
 * as logical vertex data rather than rendered pixel data, allowing them to
 * be scaled/rotated without generation loss	.
 */
public class MaglevMedium implements IMedium {
	//final List<MagLevStroke> strokes;
	//final List<MagLevFill> fills;
	List<AMagLevThing> things;
	
	final ImageWorkspace context;
	DynamicImage builtImage = null;
	private boolean isBuilt = false;
	
	public MaglevMedium( ImageWorkspace context) {
		this.context = context;
		this.things =  new ArrayList<>();
		if( context != null) {
			// TODO: unflag "isBuilt" when Workspace changes size
		}
	}
	public MaglevMedium( ImageWorkspace context, List<AMagLevThing> things) {
		this.context = context;
		this.things = new ArrayList<>( things.size());
		this.things.addAll(things);
		
		for( AMagLevThing thing : things) {
			if( thing.id == -1)
				thing.id = workingId++;
			workingId = Math.max(workingId, thing.id+1);
		}
	}
	private MaglevMedium( MaglevMedium other) {
		this.context = other.context;
		this.things = new ArrayList<>(other.things.size());
		
		for( AMagLevThing thing : other.things)
			this.things.add(thing.clone());
		
		this.isBuilt = other.isBuilt;
		this.builtImage = (other.builtImage == null) ? null : other.builtImage.deepCopy();
		this.workingId = other.workingId;
	}
	
	public ImageWorkspace getContext() {return context;}
	

	int workingId = 0;
	void addThing( AMagLevThing thing, boolean back) {
		if( thing.id == -1)
			thing.id = workingId++;
		workingId = Math.max(workingId, thing.id+1);
		if( back)
			things.add(0, thing);
		else
			things.add(thing);
	}
	void removeThing(AMagLevThing toRemove) {
		int id = toRemove.id;
		
		Iterator<AMagLevThing> it = things.iterator();
		while( it.hasNext()) {
			AMagLevThing thing = it.next();
			if( thing instanceof MagLevFill) {
				List<StrokeSegment> segments = ((MagLevFill) thing).segments;
				segments.removeIf((s) -> s.strokeId == id);
			
				if( segments.size() == 0)
					it.remove();
			}
		}
		things.remove(toRemove);
		
		unbuild();
	}
	
	// ==== Easy Junk
	@Override public int getWidth() {return (isBuilt) ? builtImage.getWidth() : context.getWidth(); }
	@Override public int getHeight() {return (isBuilt) ? builtImage.getHeight() : context.getHeight();}

	@Override public int getDynamicX() { return (isBuilt) ? builtImage.getXOffset() : 0; }
	@Override public int getDynamicY() { return (isBuilt) ? builtImage.getYOffset() : 0; }

	@Override public ABuiltMediumData build(BuildingMediumData building) {return new MaglevBuiltData(building);}
	@Override public IImageDrawer getImageDrawer(BuildingMediumData building) 
		{return new MaglevImageDrawer(this, building);}

	@Override public IMedium dupe() {return new MaglevMedium(this);}
	@Override public IMedium copyForSaving() {return new MaglevMedium(this);}
	@Override public InternalImageTypes getType() {return InternalImageTypes.MAGLEV;}

	@Override public IImage readOnlyAccess() {
		Build();
		return builtImage.getBase();
	}
	
	@Override
	public void flush() {
		if( builtImage != null)
			builtImage.flush();
		builtImage = null;
		isBuilt = false;
		//this.strokes.clear();
	}


	// ==== Hard Junk
	void splitStroke( int strokeId, float[] points) {
		int index = -1;
		for( int i=0; i < things.size(); ++i) {
			if( things.get(i).id == strokeId) {
				index = i;
				break;
			}
		}
		
		MagLevStroke stroke = (MagLevStroke)things.get(index);
		things.remove(index);
		IndexedDrawPoints direct = stroke.getDirect();
		
		
		// Step 1: Split the Stroke
		MagLevStroke[] addedStrokes = new MagLevStroke[points.length/2+1];
		boolean in = true;
		float start = 0;
		for( int i=0; i<points.length || in; ++i) {
			if( in) {
				if( (i== points.length && start != stroke.states.length-1 ) || (i < points.length && points[i] != start)) {
					ArrayList<PenState> states = new ArrayList<>();
					if( start != Math.round(start)) {
						int t = (int)Math.floor( direct.getNearIndex(start));
						
						states.add(new PenState(direct.x[t], direct.y[t], direct.w[t]));
					}
					
					if( points.length > i) {
						float endIndex = points[i];
						
						for( int c= (int)Math.ceil(start); c < endIndex; ++c) 
							states.add(stroke.states[c]);
						
						if( endIndex != Math.round(endIndex)) {
							int t = (int)Math.ceil( direct.getNearIndex(endIndex));
							
							states.add(new PenState(direct.x[t], direct.y[t], direct.w[t]));
						}
					}
					else {
						for( int c= (int)Math.ceil(start); c < stroke.states.length; ++c) 
							states.add(stroke.states[c]);
					}
					
					MagLevStroke newStroke =  new MagLevStroke(states.toArray(new PenState[0]), stroke.params);
					((AMagLevThing)newStroke).id = workingId++;
					things.add(index,newStroke);
					
					addedStrokes[i/2] = newStroke;
				}
			}
			else {
				start = points[i];
			}
			in = !in;
		}
		
		// Step 2: Re-map the Fills
		for( AMagLevThing thing : things) {
			if( thing instanceof MagLevFill) {
				MagLevFill fill = (MagLevFill)thing;
				fill.segments.removeIf( (ss) -> ss.strokeId == stroke.getId());
				// TODO: Make better
			}
		}
	}
	
	
	boolean building = false;
	void Build() {
		if( !building) {
			building = true;
			if( !isBuilt) {
				builtImage = new DynamicImage(context, HybridHelper.createNillImage(), 0, 0);
				
				builtImage.doOnGC( (gc) -> {
					ABuiltMediumData built = this.build(new BuildingMediumData(context.getHandleFor(this), 0, 0));
					SelectionMask mask = null;
					for( AMagLevThing thing : things)
						thing.draw( built, mask, gc, this);
				}, new MatTrans());
			}
			isBuilt = true;
			building = false;
		}
	}
	void unbuild() {
		if( this.isBuilt) {
			this.isBuilt = false;
			builtImage.flush();
			builtImage = null;
		}
	}
	
	public void contortBones(BaseBone bone, Interpolator2D state) {
		this.things = BoneContorter.contortBones(this.things, bone, state);
		unbuild();
		Build();
		//context.getHandleFor(this).refresh();
	}
	
	/** Be careful with these things, they can break. */
	public List<AMagLevThing> getThings() {
		return new ArrayList<AMagLevThing>(things);
	}
	

	public class MaglevBuiltData extends ABuiltMediumData {
		MatTrans trans;
		//final int box;
		//final int boy;
		
		public MaglevBuiltData(BuildingMediumData building) 
		{
			super(building.handle);
			this.trans = building.trans;
		}

		@Override public int getWidth() {return context.getWidth(); }
		@Override public int getHeight() {return context.getHeight();}

		@Override public Vec2i convert(Vec2i p) {return p;}
		@Override public Vec2 convert(Vec2 p) {return p;}

		@Override public Rect getBounds() 
		{
			// TODO
			return new Rect( (int)trans.getM02(), (int)trans.getM12(), context.getWidth(), context.getHeight());
		}
		@Override public void draw(GraphicsContext gc) 
		{
			MatTrans oldTrans = gc.getTransform();
			gc.preTransform(trans);
			gc.drawImage(builtImage.getBase(), builtImage.getXOffset(), builtImage.getYOffset());
			gc.setTransform(oldTrans);
		}
		@Override public void drawBorder(GraphicsContext gc) 
		{
			MatTrans oldTrans = gc.getTransform();
			gc.preTransform(trans);
			gc.drawRect(0, 0, context.getWidth(), context.getHeight());
			gc.setTransform(oldTrans);
		}
		@Override public MatTrans getCompositeTransform() 
			{return new MatTrans(trans);}
		@Override public MatTrans getScreenToImageTransform() 
		{
			try {
				return trans.createInverse();
			} catch (NoninvertableException e) {
				return new MatTrans();
			}
		}

		// Counter-intuitively, checking in and checking out of a MaglevInternalImage
		//	can be a thing that makes sense to do as the StrokeEngine uses it for the
		//	behavior we want.
		@Override
		protected void _doOnGC(DoerOnGC doer) {
			builtImage.doOnGC(doer, trans);
		}

		@Override
		protected void _doOnRaw(DoerOnRaw doer) {
			builtImage.doOnRaw(doer, trans);
		}
	}


	public AMagLevThing getThingById(int Id) {
		for( AMagLevThing thing : things) {
			if( thing.id == Id)
				return thing;
		}
		return null;
	}
}

File: src/spirite/base/image_data/mediums/maglev/MaglevMedium.kt
package spirite.base.image_data.mediums.maglev

import spirite.base.graphics.DynamicImage
import spirite.base.graphics.IImage
import spirite.base.image_data.ImageWorkspace
import spirite.base.image_data.ImageWorkspace.BuildingMediumData
import spirite.base.image_data.layers.puppet.BasePuppet.BaseBone
import spirite.base.image_data.mediums.BuiltMediumData
import spirite.base.image_data.mediums.DoerOnGC
import spirite.base.image_data.mediums.DoerOnRaw
import spirite.base.image_data.mediums.IMedium
import spirite.base.image_data.mediums.drawer.IImageDrawer
import spirite.base.image_data.mediums.maglev.parts.MagLevFill
import spirite.base.image_data.mediums.maglev.parts.MagLevStroke
import spirite.base.image_data.selection.SelectionMask
import spirite.base.pen.PenTraits.PenState
import spirite.base.util.interpolation.Interpolator2D
import spirite.base.util.linear.MatTrans
import spirite.hybrid.HybridHelper
import java.util.*

/**
 * A Maglev Internal Image is an image that floats just above the surface,
 * not quite planting its feat in the ground.  Essentially it is a kind of
 * Scalable Vector Image, storing all the different stroke and fill actions
 * as logical vertex data rather than rendered pixel data, allowing them to
 * be scaled/rotated without generation loss	.
 */
class MaglevMedium  (
        val context: ImageWorkspace,
        things: List<AMagLevThing>? = null
): IMedium
{
    init {
        things?.forEach {
            if( it.id == -1)
                it.id = workingId++
            workingId = Math.max(workingId, it.id+1)
        }
    }

    //internal
    var things = things?.toMutableList() ?: mutableListOf()

    //internal
    var builtImage: DynamicImage? = null
    private var isBuilt = false

    private var workingId = 0
    var building = false; private set

    private constructor(other: MaglevMedium) : this( other.context, null) {
        this.things = MutableList( other.things.size, { other.things.get(it).clone()})

        this.isBuilt = other.isBuilt
        this.builtImage = if (other.builtImage == null) null else other.builtImage!!.deepCopy()
        this.workingId = other.workingId
    }

    //internal
    fun addThing(thing: AMagLevThing, back: Boolean) {
        if (thing.id == -1)
            thing.id = workingId++
        workingId = Math.max(workingId, thing.id + 1)
        if (back)
            things.add(0, thing)
        else
            things.add(thing)
    }

    //internal
    fun removeThing(toRemove: AMagLevThing) {
        val id = toRemove.id

        val it = things.iterator()
        while (it.hasNext()) {
            val thing = it.next()
            if (thing is MagLevFill) {
                val segments = thing.segments
                segments.removeIf { s -> s.strokeId == id }

                if (segments.size == 0)
                    it.remove()
            }
        }
        things.remove(toRemove)

        unbuild()
    }

    override val width: Int get() {
        Build()
        return if (isBuilt) builtImage!!.width else context.width
    }

    override val height: Int get() {
        Build()
        return if (isBuilt) builtImage!!.height else context.height
    }

    override val dynamicX: Int get() {
        Build()
        return builtImage!!.xOffset
    }
    override val dynamicY: Int get() {
        Build()
        return if (isBuilt) builtImage!!.yOffset else 0
    }
    override val type = IMedium.InternalImageTypes.MAGLEV

    override fun build(building: BuildingMediumData): BuiltMediumData {
        return MaglevBuiltData(building)
    }

    override fun getImageDrawer(building: BuildingMediumData): IImageDrawer {
        return MaglevImageDrawer(this, building)
    }

    override fun dupe(): IMedium {
        return MaglevMedium(this)
    }

    override fun copyForSaving(): IMedium {
        return MaglevMedium(this)
    }


    override fun readOnlyAccess(): IImage {
        Build()
        return builtImage!!.base
    }

    override fun flush() {
        builtImage?.flush()
        builtImage = null
        isBuilt = false
    }


    // ==== Hard Junk
    //internal
    fun splitStroke(strokeId: Int, points: FloatArray) {
        val index = things.indices.first { things[it].id == strokeId }
        val stroke = things[index] as MagLevStroke
        val direct = stroke.direct

        things.removeAt(index)


        // Step 1: Split the Stroke
        val addedStrokes = arrayOfNulls<MagLevStroke>(points.size / 2 + 1)
        var inStroke = true
        var start = 0f
        var i = 0
        while (i < points.size || inStroke) {
            if (inStroke) {
                if (i == points.size && start != (stroke.states.size - 1).toFloat() || i < points.size && points[i] != start) {
                    val states = ArrayList<PenState>()
                    if (start != Math.round(start).toFloat()) {
                        val t = Math.floor(direct.getNearIndex(start).toDouble()).toInt()

                        states.add(PenState(direct.x[t], direct.y[t], direct.w[t]))
                    }

                    if (points.size > i) {
                        val endIndex = points[i]

                        var c = Math.ceil(start.toDouble()).toInt()
                        while (c < endIndex) {
                            states.add(stroke.states[c])
                            ++c
                        }

                        if (endIndex != Math.round(endIndex).toFloat()) {
                            val t = Math.ceil(direct.getNearIndex(endIndex).toDouble()).toInt()

                            states.add(PenState(direct.x[t], direct.y[t], direct.w[t]))
                        }
                    } else {
                        for (c in Math.ceil(start.toDouble()).toInt() until stroke.states.size)
                            states.add(stroke.states[c])
                    }

                    val newStroke = MagLevStroke(states.toTypedArray(), stroke.params)
                    (newStroke as AMagLevThing).id = workingId++
                    things.add(index, newStroke)

                    addedStrokes[i / 2] = newStroke
                }
            } else {
                start = points[i]
            }
            inStroke = !inStroke
            ++i
        }

        // Step 2: Re-map the Fills (todo)
        for (thing in things) {
            (thing as? MagLevFill)?.segments?.removeIf { ss -> ss.strokeId == stroke.id }
        }
    }

    //internal
    fun Build() {
        if (!building) {
            building = true
            if (!isBuilt) {
                val dyn = DynamicImage(context, HybridHelper.createNillImage(), 0, 0)

                dyn.doOnGC({ gc ->
                    val built = this.build(BuildingMediumData(context.getHandleFor(this), 0, 0))
                    val mask: SelectionMask? = null
                    for (thing in things)
                        thing.draw(built, mask, gc, this)
                }, MatTrans())
                builtImage = dyn
            }
            isBuilt = true
            building = false
        }
    }

    //internal
    fun unbuild() {
        if (this.isBuilt) {
            this.isBuilt = false
            builtImage!!.flush()
            builtImage = null
        }
    }

    fun contortBones(bone: BaseBone, state: Interpolator2D) {
        this.things = contortBones(this.things, bone, state)
        unbuild()
        Build()
        //context.getHandleFor(this).refresh();
    }

    /** Be careful with these things, they can break.  */
    fun getThings2(): List<AMagLevThing> {
        return ArrayList(things)
    }


    inner class MaglevBuiltData(building: BuildingMediumData) : BuiltMediumData(building) {
        init {
            Build()
        }

        override val _sourceToComposite: MatTrans get() = MatTrans()
        override val _screenToSource: MatTrans by lazy {
            val strans = MatTrans(trans)
            strans.preTranslate(builtImage!!.xOffset.toFloat(), builtImage!!.yOffset.toFloat())
            strans
        }

        override val compositeWidth: Int get() = context.width
        override val compositeHeight: Int get() = context.height

        override val sourceWidth: Int = builtImage!!.width
        override val sourceHeight: Int = builtImage!!.height

        override fun _doOnGC(doer: DoerOnGC) {
            builtImage!!.doOnGC(doer, _screenToSource)
        }

        override fun _doOnRaw(doer: DoerOnRaw) {
            builtImage!!.doOnRaw(doer, _screenToSource)
        }
        /*
        internal var trans: MatTrans = building.trans

        override fun getWidth(): Int {
            return context.width
        }

        override fun getHeight(): Int {
            return context.height
        }

        override fun convert(p: Vec2): Vec2 {
            return p
        }

        override fun getBounds(): Rect {
            // TODO
            return Rect(trans.m02.toInt(), trans.m12.toInt(), context.width, context.height)
        }

        override fun draw(gc: GraphicsContext) {
            val oldTrans = gc.transform
            gc.preTransform(trans)
            gc.drawImage(builtImage!!.base, builtImage!!.xOffset, builtImage!!.yOffset)
            gc.transform = oldTrans
        }

        override fun drawBorder(gc: GraphicsContext) {
            val oldTrans = gc.transform
            gc.preTransform(trans)
            gc.drawRect(0, 0, context.width, context.height)
            gc.transform = oldTrans
        }

        override fun getCompositeTransform(): MatTrans {
            return MatTrans(trans)
        }

        override fun getScreenToImageTransform(): MatTrans {
            try {
                return trans.createInverse()
            } catch (e: NoninvertableException) {
                return MatTrans()
            }

        }

        // Counter-intuitively, checking in and checking out of a MaglevInternalImage
        //	can be a thing that makes sense to do as the StrokeEngine uses it for the
        //	behavior we want.
        override fun _doOnGC(doer: DoerOnGC) {
            builtImage!!.doOnGC(doer, trans)
        }

        override fun _doOnRaw(doer: DoerOnRaw) {
            builtImage!!.doOnRaw(doer, trans)
        }*/
    }


    fun getThingById(Id: Int): AMagLevThing? {
        for (thing in things) {
            if (thing.id == Id)
                return thing
        }
        return null
    }
}
